### PR TITLE
http: json_to_metadata filter

### PR DESCRIPTION
**Pull request by JuniorHsu**
Commit Message:

This filter is configured with rules that will be matched against requests. Each rule has either a series of selectors and can be triggered either when the json attribute is present or missing.

When a rule is triggered, dynamic metadata will be added based on the configuration of the rule. If present, it’s value is extracted and used along with the specified key as metadata. If missing, on missing case is triggered and the value specified is used for adding metadata.

The metadata can then be used for load balancing decisions, consumed from logs, etc.

A typical use case for this filter is to dynamically match a specified json attribute of requests with rate limit. For this, a given json payload attribute’s value would be extracted and attached to the request as dynamic metadata which would then be used to match a rate limit action on metadata.

The Json to metadata filter stops iterating to next filter until we have the whole payload to extract the Json attributes or encounter error.

Additional Description:
Risk Level: low
Testing: unit
Docs Changes: new doc for filter
Release Notes: current.yaml
[Optional Fixes #Issue] #27645

### CODE REVIEW

File: CODEOWNERS
Comment (yanavlasov) [inline CODEOWNERS:286]: A new extension needs a maintainer sponsor to be added to the main repository. Please find a maintainer who will take ownership together with you. Otherwise you can add it to the contrib directory.

Comment (JuniorHsu) [reply CODEOWNERS]: updated

File: api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto
Comment (markdroth) [inline api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto:48]: What does the term "attribute" refer to here?  Does it mean the value of the JSON key?  If so, I suggest using a phrase like "the value of the JSON key" instead.

Same thing throughout.

File: api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto
Comment (markdroth) [inline api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto:98]: Instead of building your own matching API, have you considered using the [xDS generic matching API](https://github.com/cncf/xds/blob/main/xds/type/matcher/v3/matcher.proto)?  See the original design [here](https://docs.google.com/document/d/1G4g-6q0IArz_ERgqixzZCM0-wbexFYuBUvT9hkV7QRU/edit).

File: source/extensions/filters/http/json_to_metadata/filter.h
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.h:41]: consider just making this a struct with the keys_ vector const. Then we could get rid of getters and the entire object would be "const"

File: source/extensions/filters/http/json_to_metadata/filter.h
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.h:54]: Can the classes here be documented some more?

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.cc:66]: pass string by const reference?

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.cc:355]: Why not use the `requestBufferLimitBytes` to set the decoder_callbacks_ buffer limit to be independent of the "default behavior" of when Envoy stops buffering ? Also it seems like we will fail open, should this be a local reply?

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.cc:184]: consider putting the various helper function objects in an anonymous namespace at the start of the file

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.cc:324]: what exactly is a reported request?

**Review by KBaichoo — COMMENTED** [review]
Here's a first pass, thanks for working on this

/wait

File: api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto
Comment (JuniorHsu) [reply api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto:48]: thanks for the suggestion! this wording bothers me a lot

File: api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto
Comment (JuniorHsu) [reply api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto:98]: We haven't walked through various matchers but seek for MVP while we implemented this. I just did a quick check the xds matching API that current utilities are around http requests? Json matchers might have very different natures and we are happy on current design as it fits the needs. Does it make sense?

File: api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto
Comment (JuniorHsu) [reply api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto:103]: yeah although this is not managed by the filter.

File: api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto
Comment (JuniorHsu) [reply api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto:115]: do you mean to introduce a new struct like
```
message Rules {
  repeated Rule rules = 1 [(validate.rules).repeated = {min_items: 1}];
  google.protobuf.UInt32Value limit_bytes = 2;
  ...
}


Rules request_rules = 1;
```

so in the future we can add `Rules response_rules = 2;` ?

File: api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto
Comment (JuniorHsu) [reply api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto:48]: most of the cases "attribute" means the key-value pair of the desired target.
However "attribute value" should be "the value of the JSON key"

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (JuniorHsu) [reply source/extensions/filters/http/json_to_metadata/filter.cc:324]: hmmm. it means the request body is parsed and set the corresponding metadata. should be `is_request_handled_` ?

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (JuniorHsu) [reply source/extensions/filters/http/json_to_metadata/filter.cc:355]: It's a good point and I didn't notice we can set that.
By reading the comments, seems not recommended to shrink the limit but not a hard limitation.
I agree that we shouldn't re-invent the wheel. So probably just set the bytes in `decodeHeaders` or even constructor?

https://github.com/envoyproxy/envoy/blob/ce9c61fe760fafa6d3033fc6bc45a013d37a8b5e/envoy/http/filter.h#L682-L693

File: source/extensions/filters/http/json_to_metadata/filter.h
Comment (JuniorHsu) [reply source/extensions/filters/http/json_to_metadata/filter.h:39]: which exact style clause you're referring to? Also many prior art exists.


```
source/extensions/filters/http/file_system_buffer/filter_config.h:22:using ProtoStreamConfig = envoy::extensions::filters::http::file_system_buffer::v3::StreamConfig;
source/extensions/filters/http/header_mutation/header_mutation.h:21:using ProtoConfig = envoy::extensions::filters::http::header_mutation::v3::HeaderMutation;
source/extensions/filters/http/header_mutation/header_mutation.h:24:using MutationsProto = envoy::extensions::filters::http::header_mutation::v3::Mutations;
source/extensions/filters/http/stateful_session/stateful_session.h:22:using ProtoConfig = envoy::extensions::filters::http::stateful_session::v3::StatefulSession;
source/extensions/filters/http/tap/tap_config_impl.h:47:  using HttpStreamedTraceSegment = envoy::data::tap::v3::HttpStreamedTraceSegment;
source/extensions/filters/http/tap/tap_config_impl.h:48:  using MutableBodyChunk = envoy::data::tap::v3::Body* (HttpStreamedTraceSegment::*)();
source/extensions/filters/http/tap/tap_config_impl.h:49:  using HttpBufferedTrace = envoy::data::tap::v3::HttpBufferedTrace;
source/extensions/filters/http/tap/tap_config_impl.h:50:  using MutableMessage = envoy::data::tap::v3::HttpBufferedTrace::Message* (HttpBufferedTrace::*)();
source/extensions/filters/network/dubbo_proxy/config.h:41:  using DubboProxyConfig = envoy::extensions::filters::network::dubbo_proxy::v3::DubboProxy;
source/extensions/load_balancing_policies/maglev/config.h:16:using MaglevLbProto = envoy::extensions::load_balancing_policies::maglev::v3::Maglev;
source/extensions/load_balancing_policies/random/config.h:15:using RandomLbProto = envoy::extensions::load_balancing_policies::random::v3::Random;
source/extensions/load_balancing_policies/ring_hash/config.h:16:using RingHashLbProto = envoy::extensions::load_balancing_policies::ring_hash::v3::RingHash;
source/extensions/load_balancing_policies/ring_hash/ring_hash_lb.h:57:  using HashFunction = envoy::config::cluster::v3::Cluster::RingHashLbConfig::HashFunction;
source/extensions/load_balancing_policies/round_robin/config.h:15:using RoundRobinLbProto = envoy::extensions::load_balancing_policies::round_robin::v3::RoundRobin;
source/extensions/load_balancing_policies/subset/subset_lb.h:115:  using SubsetFallbackPolicy = envoy::config::cluster::v3::Cluster::LbSubsetConfig::
source/extensions/stat_sinks/metrics_service/grpc_metrics_service_impl.h:26:using HistogramEmitMode = envoy::config::metrics::v3::HistogramEmitMode;

```

File: source/extensions/filters/http/json_to_metadata/filter.h
Comment (JuniorHsu) [reply source/extensions/filters/http/json_to_metadata/filter.h:41]: yep it's 
```
using Rules = std::vector<Rule>;

Rules request_rules_;
```
in the member so i'll introduce a helper function to make `const Rules request_rules_`

File: source/extensions/filters/http/json_to_metadata/filter.h
Comment (JuniorHsu) [reply source/extensions/filters/http/json_to_metadata/filter.h:54]: Do you want more doc for member functions or just classes?

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.cc:90]: note this is not using the underlying connection buffer value -- just the connection buffer size default of 1MB. e.g. say a deployment sets that buffer to 2MB, this would just be 1MB if unset. It should either be tied to that value or, I think (better) be independent of it.

For the latter that means that we can have this be some value independent of connection buffer, defaulting to using either 1MB or the HTTP filter chain's default sizes -- though this is protocol dependent.

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (KBaichoo) [reply source/extensions/filters/http/json_to_metadata/filter.cc:370]: Yea I think we should leverage setting if this filters requests a higher limit than is present. If another filter has requested a higher limit, we must decide whether we want this to fail and send a local reply, etc.

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.cc:369]: We should also take into consideration the decoding buffer passed to the function, it's possible that that buffer != `decodingBuffer()` and we can miss triggering this if condition (especially if we leverage setting the decoding buffer limit.

For example, if 100B is our limit:
```
decodeData(80 byte input, no end stream) ; decodingBuffer() size: 0
decodeData(40 byte input, no end stream); decodingBuffer() size: 80
<-- we won't get another call of decodeData as when we tell it to buffer the 40, we should have 120 now in decodingBuffer 
```

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.cc:176]: No need to do the double find, we can just do 
`auto& keyval = struct_map[meta_namespace]` 

which should insert the default ctor if key not exist or provide you the value if it does.

File: test/extensions/filters/http/json_to_metadata/filter_test.cc
Comment (KBaichoo) [inline test/extensions/filters/http/json_to_metadata/filter_test.cc:1]: Recommend adding some integration test using the filter as it'll allow us to test the filter usage as when it's in the filter chain

**Review by KBaichoo — COMMENTED** [review]
Here's another pass

/wait

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (JuniorHsu) [reply source/extensions/filters/http/json_to_metadata/filter.cc:339]: I'll use `request_processing_finished_` if you don't mind

**Comment by JuniorHsu** [comment]
Handled all comments modulo the buffering and integration test.

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (JuniorHsu) [reply source/extensions/filters/http/json_to_metadata/filter.cc:362]: I'm not 100% sure but while we get `end_stream == true` the `decodingBuffer()` will be the full buffer? I was guessing the `data` of the parameter `Filter::decodeData(Buffer::Instance& data` is a *buffer_view* on `decodingBuffer()` . It would be hard to manage the status if `buffer != decodingBuffer()` I'm open to suggestion on how to make it right.

On the other hand, should we suggest filter users to set `per_connection_buffer_limit_bytes` and remove the `bufferLimitBytes` support in the filter, since the buffer limit is to avoid oom and connection level limitation is probably better. Although one big request could punish other light-weight request, we have overload manager to mitigate that.

wdyt @KBaichoo ?

File: test/extensions/filters/http/json_to_metadata/filter_test.cc
Comment (JuniorHsu) [reply test/extensions/filters/http/json_to_metadata/filter_test.cc:1]: done. While I wrote my first filter and I have a bug in the factory code so which is not able to test in config/filter_test. Thanks for suggestion!

File: source/extensions/filters/http/json_to_metadata/filter.h
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.h:77]: this seems const

File: source/extensions/filters/http/json_to_metadata/filter.h
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.h:75]: likewise as below

File: source/extensions/filters/http/json_to_metadata/filter.h
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.h:111]: can you provide insight on why we pass around the reported_flag which is just the `request_processing_finished_` flag?

**Review by KBaichoo — COMMENTED** [review]
/wait

File: source/extensions/filters/http/json_to_metadata/filter.h
Comment (JuniorHsu) [reply source/extensions/filters/http/json_to_metadata/filter.h:111]: in the future we might handle response as well, so have a `out` variable. I'll rename it to `processing_finished_flag`

File: test/extensions/filters/http/json_to_metadata/integration_test.cc
Comment (JuniorHsu) [reply test/extensions/filters/http/json_to_metadata/integration_test.cc]: that catch a real bug. thanks!

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (JuniorHsu) [reply source/extensions/filters/http/json_to_metadata/filter.cc]: remove for now

File: test/extensions/filters/http/json_to_metadata/filter_test.cc
Comment (KBaichoo) [inline test/extensions/filters/http/json_to_metadata/filter_test.cc:130]: const (here and elsewhere)

**Review by KBaichoo — COMMENTED** [review]
Thanks this is looking much better.

**Comment by KBaichoo** [comment]
/wait

Needs merge

**Comment by JuniorHsu** [comment]
Thanks for review! I'll resume the work next Thursday.

File: source/extensions/filters/http/json_to_metadata/filter.cc
Comment (JuniorHsu) [reply source/extensions/filters/http/json_to_metadata/filter.cc:236]: Oh the exception is gone after I introduced json utilities no throw version. I'll remove those legacy comments. Thanks.

File: test/extensions/filters/http/json_to_metadata/integration_test.cc
Comment (JuniorHsu) [reply test/extensions/filters/http/json_to_metadata/integration_test.cc:97]: I had hard time to understand what to test watermark in json to metadata filter.
I assume watermark is buffer limit here but this filter doesn't really do the local reply based on filter buffer limit and so really i'm testing the test framework in my latest pr instead of testing the json to metadata filter.

I don't have the best knowledge on the filter buffer limit. Therefore I need more help if the buffer limit integration test  is necessary.

**Comment by JuniorHsu** [comment]
ci error is probably unrelated. would retest tmr
```
WARNING: Download from https://github.com/unicode-org/icu/archive/release-72-1.tar.gz failed: class java.io.IOException java.net.SocketTimeoutException: Read timed out
ERROR: An error occurred during the fetch of repository 'com_github_unicode_org_icu':
   Traceback (most recent call last):
	File "/build/bazel_root/base/external/bazel_tools/tools/build_defs/repo/http.bzl", line 132, column 45, in _http_archive_impl
		download_info = ctx.download_and_extract(
Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/unicode-org/icu/archive/release-72-1.tar.gz] to /build/bazel_root/base/external/com_github_unicode_org_icu/temp17244805646648187250/release-72-1.tar.gz: java.net.SocketTimeoutException: Read timed out
ERROR: /source/WORKSPACE:13:19: fetching http_archive rule //external:com_github_unicode_org_icu: Traceback (most recent call last):
	File "/build/bazel_root/base/external/bazel_tools/tools/build_defs/repo/http.bzl", line 132, column 45, in _http_archive_impl
		download_info = ctx.download_and_extract(
Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/unicode-org/icu/archive/release-72-1.tar.gz] to /build/bazel_root/base/external/com_github_unicode_org_icu/temp17244805646648187250/release-72-1.tar.gz: java.net.SocketTimeoutException: Read timed out
```

File: test/extensions/filters/http/json_to_metadata/integration_test.cc
Comment (KBaichoo) [inline test/extensions/filters/http/json_to_metadata/integration_test.cc:146]: Test is failing in arm, and windows, looks legit:

```
[ RUN      ] Protocols/JsonToMetadataIntegrationTest.GoOverWaterMarkBeforeEndStream/IPv4_HttpDownstream_HttpUpstreamHttpParserNghttp2NoDeferredProcessing
[2023-08-04 05:43:16.081][62580][critical][backtrace] [./source/server/backtrace.h:104] Caught Segmentation fault, suspect faulting address 0xbeb39f8000000010
[2023-08-04 05:43:16.081][62580][critical][backtrace] [./source/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2023-08-04 05:43:16.081][62580][critical][backtrace] [./source/server/backtrace.h:92] Envoy version: 0/1.28.0-dev/test/RELEASE/BoringSSL
[2023-08-04 05:43:16.081][62580][critical][backtrace] [./source/server/backtrace.h:96] #0: __kernel_rt_sigreturn [0xffffb8aad78c]
[2023-08-04 05:43:16.092][62580][critical][backtrace] [./source/server/backtrace.h:96] #1: Envoy::IntegrationCodecClient::sendData() [0xa8eb80]
[2023-08-04 05:43:16.103][62580][critical][backtrace] [./source/server/backtrace.h:96] #2: Envoy::(anonymous namespace)::JsonToMetadataIntegrationTest::runTest() [0xa6f160]
[2023-08-04 05:43:16.114][62580][critical][backtrace] [./source/server/backtrace.h:96] #3: Envoy::(anonymous namespace)::JsonToMetadataIntegrationTest_GoOverWaterMarkBeforeEndStream_Test::TestBody() [0xa72538]
[2023-08-04 05:43:16.125][62580][critical][backtrace] [./source/server/backtrace.h:96] #4: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x1caa7b8]
```

File: source/extensions/filters/http/json_to_metadata/filter.h
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.h:63]: Remove defaulted ctor (compiler will provide)

File: source/extensions/filters/http/json_to_metadata/filter.h
Comment (KBaichoo) [inline source/extensions/filters/http/json_to_metadata/filter.h:93]: likewise

File: test/extensions/filters/http/json_to_metadata/integration_test.cc
Comment (KBaichoo) [reply test/extensions/filters/http/json_to_metadata/integration_test.cc:97]: ack, it might no longer be necessary since we removed the buffering limit this filter has; still the test is failing -- while we don't __need__ it , can you figure out why? We might be missing something.

**Review by KBaichoo — COMMENTED** [review]
Generally looks good besides few remaining points.

/assign-from @envoyproxy/senior-maintainers

File: test/extensions/filters/http/json_to_metadata/integration_test.cc
Comment (JuniorHsu) [reply test/extensions/filters/http/json_to_metadata/integration_test.cc:97]: reply in another thread. thanks!

File: test/extensions/filters/http/json_to_metadata/integration_test.cc
Comment (JuniorHsu) [reply test/extensions/filters/http/json_to_metadata/integration_test.cc:146]: It crashes in the testing framework here with dangling raw point `inner_encoder_`
https://github.com/envoyproxy/envoy/blob/b1ea06265c05799f4750f577090d2a33d5829fd8/source/common/http/codec_wrappers.h#L82

```
(gdb) p inner_encoder_
$4 = (Envoy::Http::RequestEncoder *) 0x2578bd512570
(gdb) info vtbl inner_encoder_
warning: can't find linker symbol for virtual table for `Envoy::Http::RequestEncoder' value
```

I suspect I abused the test framework :P 

Moreover, I substitue the prepended filter by `wait-for-whole-request-and-response-filter`
to simulate the buffer behavior and it still crashes, which kinda proves the crash is not related to
json to metadata filter. 

I'm going to remove the unnecessary test to move forward.

**Review by RyanTheOptimist — APPROVED** [review]
Looks great!

**Comment by KBaichoo** [comment]
/unassign @markdroth 

/assign @adisuissa 

For API review since Roth is out.

**Comment by JuniorHsu** [comment]
@adisuissa FWIW roth did approve [once](https://github.com/envoyproxy/envoy/pull/28394#issuecomment-1642953738) but we need a fresh stamp.  Thanks!

**Review by adisuissa — COMMENTED** [review]
Following [the approval](https://github.com/envoyproxy/envoy/pull/28394#issuecomment-1642953738) here it is again:
/lgtm api

**Pull request merged**
